packages/augur-core/src/contracts/trading/erc20proxy1155/ERC20Proxy1155Nexus.sol

```diff
@@ -0,0 +1,84 @@
+pragma solidity 0.5.15;
+
+import 'ROOT/libraries/CloneFactory.sol';
+import 'ROOT/libraries/token/IERC20.sol';
+import 'ROOT/libraries/token/IERC1155.sol';
+import 'ROOT/libraries/math/SafeMathUint256.sol';
+
+import 'ROOT/trading/erc20proxy1155/IERC20Proxy1155Nexus.sol';
+import 'ROOT/trading/erc20proxy1155/IERC20Proxy1155.sol';
+
+// To use these ERC20s, the user will have to call
+// setApprovalForAll(ERC20Proxy1155Nexus, true)
+// on the ShareToken contract themselves.
+
+contract ERC20Proxy1155Nexus is IERC20Proxy1155Nexus {
+    using SafeMathUint256 for uint256;
+
+    IERC1155 public target1155;
+    IERC20Proxy1155 proxyToClone;
+
+    // tokenId -> owner -> spender -> allowance
+    // The spender is whoever called the proxy contract, which is msg.sender in the proxy and _caller here in the nexus.
+    mapping (uint256 => mapping (address => mapping (address => uint256))) public allowances;
+    // tokenId -> proxy contract
+    mapping (uint256 => IERC20Proxy1155) public proxies;
+
+    constructor(address _proxyToClone, IERC1155 _target1155) public {
+        proxyToClone = IERC20Proxy1155(_proxyToClone);
+        target1155 = _target1155;
+    }
+
+    function newERC20(uint256 _tokenId) public returns (IERC20Proxy1155) {
+        require(proxies[_tokenId] == IERC20Proxy1155(0)); // must not have been registered yet
+        IERC20Proxy1155 _erc20 = IERC20Proxy1155(createClone(address(proxyToClone)));
+        proxies[_tokenId] = _erc20; // register
+        _erc20.initialize(this, _tokenId);
+        return _erc20;
+    }
+
+    function transfer(IERC20Proxy1155 _proxy, address _caller, address _to, uint256 _amount) public returns (bool) {
+        _transfer(_proxy, _caller, _to, _amount);
+        return true;
+    }
+
+    function transferFrom(IERC20Proxy1155 _proxy, address _caller, address _from, address _to, uint256 _amount) public returns (bool) {
+        _transfer(_proxy, _from, _to, _amount);
+        _approve(_proxy, _from, _caller, allowances[_proxy.tokenId()][_from][_caller].sub(_amount));
+        return true;
+    }
+
+    function approve(IERC20Proxy1155 _proxy, address _caller, address _spender, uint256 _amount) public returns (bool) {
+        _approve(_proxy, _caller, _spender, _amount);
+        return true;
+    }
+
+    function allowance(IERC20Proxy1155 _proxy, address _owner, address _spender) public view returns (uint256) {
+        uint256 _tokenId = _proxy.tokenId();
+        return allowances[_tokenId][_owner][_spender];
+    }
+
+    function _transfer(IERC20Proxy1155 _proxy, address _from, address _to, uint256 _amount) internal {
+        uint256 _tokenId = _proxy.tokenId();
+        _gate(_proxy, _tokenId);
+
+        bytes memory _nothing = new bytes(0);
+        target1155.safeTransferFrom(_from, _to, _tokenId, _amount, _nothing);
```

Comment:
We should actually use `unsafeTransferFrom` which doesn't even accept a data argument. The problem with `safeTransferFrom` is that while it prevents people form accidentally depositing into contracts not willing to accept them that way, it also raises a bunch of ways a malicious user can lock up a trading contract and puts a requirement on contract integrations to support a very specific standard. `unsafeTransferFrom` is ironically probably way safer

Reply:
I can do that. `unsafeTransferFrom` is just on ShareToken though so I'd need to integrate against it specifically instead of a generic 20-1155 proxy. Which is fine I just liked that it was generic.

Reply:
done